### PR TITLE
Add `WeightNorm` reparametrization

### PR DESCRIPTION
**Pull request by pxl-th**
Yet  another attempt at adding WeightNorm.
Based on different bits found in #2053 [edited!] & #1005 with tests and documentation.

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

**Comment by CarloLucibello** [comment]
This way we have duplication of the weights. 
Can we let the original layer's weights act like `v`?
Something like
```julia
struct WeightNorm{which, dims, L, G}
    layer::L
    g::G
end

(w::WeightNorm)(x) = transform(w)(x)

function transform(wn::WeightNorm{which, dims}) where {which, dims}
    ϵ = eps(eltype(wn.v))
    v = getfield(wn.layer, which)     
    n2 = sum(abs2, v; dims)
    w = @. wn.g * v / sqrt(n2 + ϵ)
    fields, ctor = Functors.functor(wn.layer)
    return ctor(merge(
        fields, NamedTuple{(which,)}((w,)),
    ))
end
```

File: src/layers/normalise.jl [hunk 1]
```diff
@@ -568,3 +568,117 @@ scale parameters, `false` otherwise.
...
+    end
+
+    g = sqrt.(sum(abs2, x; dims) .+ eps(eltype(x)))
+    x ./= g # Store `v` in the original weights.
+    WeightNorm{which, dims, L, typeof(g)}(layer, g)
+end
+
+(w::WeightNorm)(x) = transform(w)(x)
+
+function transform(wn::WeightNorm{which, dims}) where {which, dims}
```
[the remark this answers is not in the document]
Comment (pxl-th) [reply src/layers/normalise.jl]: Named it `reparametrize`, let me know if it works.

**Comment by pxl-th** [comment]
Added GPU tests as well, making a small test suite which we can exapnd in future to avoid duplication.

**Comment by CarloLucibello** [comment]
The doctest needs to be fixed. besides that looks good

**Comment by pxl-th** [comment]
All tests now pass.

**Pull request merged**